### PR TITLE
[model] feat: add glm4v

### PR DESCRIPTION
**Pull request by lambertwjh**
### What does this PR do?

Add GLM4.1V support

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/utils/dataset/rl_dataset.py [hunk 1]
```diff
@@ -305,6 +305,19 @@ def __getitem__(self, item):
                 )
             ]  # (1, 3, seq_len)
 
+        if self.processor is not None and "Glm4vImageProcessor" in self.processor.image_processor.__class__.__name__:
+            from verl.models.transformers.glm4v import get_rope_index
```
Comment (hiyouga) [inline verl/utils/dataset/rl_dataset.py:315]: Is this different from qwen2vl's mrope?

File: verl/workers/actor/dp_actor.py
Comment (hiyouga) [inline verl/workers/actor/dp_actor.py:1]: It is unacceptable to modify the dp actor file significantly. Please refer to MiniCPMo's PR and wrap it in an external function. https://github.com/volcengine/verl/pull/2212/files#diff-e06e73d3a7775a502b7aea91103e7911f6597eb48e4b898db558766cdd41daf9

File: verl/utils/dataset/rl_dataset.py (hunk 1 above)
Comment (lambertwjh) [reply verl/utils/dataset/rl_dataset.py:315]: yep

File: verl/workers/actor/dp_actor.py
Comment (lambertwjh) [reply verl/workers/actor/dp_actor.py:1]: copy

**Comment by kfallah** [comment]
Hi, nice PR! I found that this change was also needed for me to get glm4v working with PEFT: https://github.com/volcengine/verl/commit/1e353c33b23fbf6a4e874fd8668316c49c200eab. Feel free to cherry-pick it.

File: verl/utils/dataset/rl_dataset.py (hunk 1 above)
Comment (hiyouga) [reply verl/utils/dataset/rl_dataset.py:315]: got it

File: verl/models/mcore/registry.py [hunk 2]
```diff
@@ -35,7 +36,6 @@
 )
 from .model_forward import (
     gptmodel_forward,
-    gptmodel_forward_qwen2_5_vl,
```
Comment (hiyouga) [inline verl/models/mcore/registry.py:38]: please do not modify this line

**Comment by lambertwjh** [comment]
@hiyouga check plz

**Comment by zRzRzRzRzRzRzR** [comment]
Some tests are still being updated, so try to minimize changes.

File: verl/models/mcore/registry.py [hunk 3]
```diff
@@ -86,7 +86,6 @@ class SupportedModel(Enum):
     SupportedModel.LLAMA4: hf_to_mcore_config_llama4,
     SupportedModel.QWEN3: hf_to_mcore_config_dense,
     SupportedModel.QWEN3_MOE: hf_to_mcore_config_qwen3moe,
-    SupportedModel.QWEN2_5_VL: hf_to_mcore_config_qwen2_5_vl,
```
Comment (ISEEKYAN) [inline verl/models/mcore/registry.py:89]: why remove qwen2_5_vl?

Comment (zRzRzRzRzRzRzR) [reply verl/models/mcore/registry.py:89]: These two code are repetitive.
There is a sentence with the same content above.

**Comment by zRzRzRzRzRzRzR** [comment]
As GLM
```
def extract_multi_modal_inputs(
    batch_data: list[dict[str, torch.Tensor]],
    indices: Optional[list[int]] = None,
) -> dict[str, torch.Tensor | list[torch.Tensor]]:
    """
    Extract and process multi-modal inputs from a batch.

    Args:
        batch_data (list[dict[str, torch.Tensor]]): The batch containing potential multi-modal inputs
        indices (Optional[list[int]]): If provided, only extract inputs at these indices

    Returns:
        dict[str, torch.Tensor | list[torch.Tensor]]: Processed multi-modal inputs ready for model consumption

    """
    multi_modal_inputs = {}
    multi_modal_inputs_collected = {}
    has_image_bound = False

    selected_batch_data = batch_data
    if indices is not None:
        selected_batch_data = [batch_data[i] for i in indices if i < len(batch_data)]

    for inputs in selected_batch_data:
        if "image_bound" in inputs:
            has_image_bound = True
        for key, value in inputs.items():
            if value is not None:
                if key not in multi_modal_inputs_collected:
                    multi_modal_inputs_collected[key] = []
                multi_modal_inputs_collected[key].append(value)

    for key, values in multi_modal_inputs_collected.items():
        if has_image_bound:  # minicpm-o logic
            multi_modal_inputs[key] = values
        else:
            all_keys = set().union(*(set(d.keys()) for d in selected_batch_data))
            for key in all_keys:
                vals = []
                if key in ("pixel_values", "pixel_values_videos"):
                    for d in selected_batch_data:
                        if key in d and d[key] is not None:
                            v = d[key]
                            vals.append(v.squeeze(0) if v.ndim == 3 else v)
                    if len(vals) > 0:
                        multi_modal_inputs[key] = torch.cat(vals, dim=0)
                elif key in ("image_grid_thw", "video_grid_thw", "second_per_grid_ts"):
                    for d in selected_batch_data:
                        if key in d and d[key] is not None:
                            vals.append(d[key])
                    if len(vals) > 0:
                        multi_modal_inputs[key] = torch.cat(vals, dim=0)
                else:
                    for d in selected_batch_data:
                        if key in d and d[key] is not None:
                            vals.append(d[key])
                    if len(vals) > 0:
                        multi_modal_inputs[key] = torch.cat(vals, dim=0)

    return multi_modal_inputs
```
should changed likes this and working.

File: verl/workers/actor/dp_actor.py [hunk 4]
```diff
@@ -491,4 +491,4 @@ def update_policy(self, data: DataProto):
                 mini_batch_metrics = {"actor/grad_norm": grad_norm.detach().item()}
                 append_to_dict(metrics, mini_batch_metrics)
         self.actor_optimizer.zero_grad()
-        return metrics
+        return metrics
```
Comment (hiyouga) [inline verl/workers/actor/dp_actor.py]: Could you fix the end of file

File: verl/workers/rollout/vllm_rollout/vllm_rollout_spmd.py [hunk 5]
```diff
@@ -401,6 +401,21 @@ def generate_sequences(self, prompts: DataProto, **kwargs) -> DataProto:
...
+            for i in range(batch_size):
+                response_text = self.tokenizer.decode(response[i], skip_special_tokens=True)
+                think_data = self._extract_think_content(response_text)
+                think_info.append(think_data)
+
+            think_mask = self._create_think_mask(
+                response, self.think_token_id, self.think_end_token_id
+            )
+
+            batch["think_mask"] = think_mask
```
Comment (hiyouga) [inline verl/workers/rollout/vllm_rollout/vllm_rollout_spmd.py]: where did the think mask use?

Comment (zRzRzRzRzRzRzR) [reply verl/workers/rollout/vllm_rollout/vllm_rollout_spmd.py]: remove it

**Comment by zRzRzRzRzRzRzR** [comment]
All changes have been completed.

File: verl/models/transformers/monkey_patch.py [hunk 6]
```diff
@@ -323,3 +341,40 @@ def state_dict(self, *args, **kwargs):
...
+        if 'position_ids' in kwargs:
+            kwargs['position_ids'] = None
+        elif len(args) >= 3:
+            args = list(args)
+            args[2] = None
+            args = tuple(args)
+    
+    return original_preprocess_mask_arguments(*args, **kwargs)
+
+transformers.masking_utils.create_causal_mask = _intercepted_create_causal_mask
```
Comment (hiyouga) [inline verl/models/transformers/monkey_patch.py]: Please restrict this patch to glm4v

**Comment by Z-MU-Z** [comment]
Hi! Thanks for merging PR #3291 that adds GLM-4.1V support. Do you have plans to support GLM-4.5V (with both thinking and non-thinking modes)?
